redhat-developer/build
# Move the steps resources definition into the strategies

For issue #239 

This PR includes many changes, most of them not critical, mainly the following:
- refactor code/docs/samples
- add more docs on how to define resources via the strategies(cluster-scope or namespaced-scope)
- ensure unit-tests are passing
- ensure the `generate_taskrun.go` can process the resources properly.

Based on the commits comments, these are the changes:

Remove references to the Resources object:
- From the Build/BuildRun API
- From the CustomResourceDefinition´s, via `operator-sdk generate crds`
- Regenerate deepcopy.go files, via `operator-sdk generate k8s`

Adapt unit-tests, the generate_taskrun_test.go required a big refactoring to achieve the following:
- Generate Build/BuildRun/BuildStrategies from YAML files definition, rather than
  from k8s objects. This is more accurate, while then we can properly define the
  step resources of any strategy on the same way how we would be doing on a k8s
  cluster, e.g. via a YAML file.
- For the above, I required to generate three new files that can allow us to host
  YAML declarations for each type of CRD. I find this more clean and readable.
- Update the catalog.go with some helper functions to parse the above YAMLs, this
  requires a new library, while parsing the strategy steps with `go-yaml/yaml` will
  not work easily. We need a way to move from YAML to JSON, and from JSON to an object,
  therefore the usage of sigs.k8s.io/yaml.
- Ensure all existing unit-tests continue working.

Update resources definition on our test samples for e2e:
- Remove any spec.resources definition in the Build/BuildRun examples
- Move all resources definition into all the BuildStrategies/ClusterBuildStrategy
  examples

Enhance docs:
- Add docs around how to define resources via the strategies
- Remove references to the old spec.resources approach via the Build
- Add docs on how Tekton deals with each step/container resources
